ApiData.ts

```
@@ -0,0 +1,250 @@
+/** Debounce utility function to limit function execution frequency */
+function debounce<F extends (...args: any) => any>(func: F, waitFor: number) {
+	let timeout: number;
+
+	return (...args: Parameters<F>): Promise<ReturnType<F>> => {
+		clearTimeout(timeout);
+
+		return new Promise((resolve) => {
+			timeout = setTimeout(() => {
+				resolve(func(...args));
+			}, waitFor);
+		});
+	};
+}
+
+/** Constants for grid calculation
+* GRID_RATIO represents the ratio of the grid's height to the window's height.
+*/
+const GRID_RATIO = 9 / 20;
+const ROW_HEIGHT = 16;
+
+/** manage data and settings on the grid */
+class ApiData {
+
+	pageSize: number;
+	currentPage: number = 1;
+	data: GridData[] = [];
+	totalItems: number = 0;
+	columnNames: ColumnName[] = [];
+	maxGridHeight: number = 0;
+	firstVal: number = 0;
+	lastVal: number = -1;
+	maxRange: number;
+
+	constructor(pageSize: number) {
+		this.pageSize = pageSize;
+		this.maxRange = 0;
+	}
+
+	/** Initialize method to set up the grid */
+	initialize(): Promise<void> {
+		this.adjustGridHeight();
+		return this.recordCount()
+			.then(() => this.fetchColumns())
+			.then(() => this.fetchAndDisplayRecords())
+			.then(() => this.setupControls());
+	}
+
+	/** Fetch total record count from the server,fetches data from an API and populates class properties */
+	recordCount(): Promise<void> {
+		return this.fetchNumData('http://localhost:2050/recordCount')
+			.then((response: number) => {
+				this.totalItems = response;
+				this.maxRange = this.totalItems - 1;
+			})
+			.catch(error => {
+				console.error('Failed to fetch record count:', error);
+				alert('Failed to fetch record count.');
+				throw error;
```

???

You handle the error... but then you throw it anyway? Rather just throw an error here, and handle the error as late as possible.